uswds/uswds
# Form Inputs: Normalize disabled forms styles

[Preview →](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/cm-disabled-forms-styles/)

Resolves #5051
Includes work from: 
#5004 
#5128 
#5146 

## Description

Creates and normalizes disabled styles across USWDS forms.

## Breaking change
:warning: This is potentially a breaking change.

Existing `usa-combo-box`, `usa-date-picker`, `usa-file-input` components with `aria-disabled="true"` may be affected now that the javascript for each component looks for that attribute during initialization

While we advise against using the usa-button--disabled class, this would break markdown for devs who have implemented it

## Related PRs

[USWDS Site change log PR →](https://github.com/uswds/uswds-site/pull/2089)

## The Problem

Currently, Chrome and Edge support disabled styles for forms. Both implement a pretty minimal opacity shift to disabled elements to help users understand an element is disabled. The shift is pretty minor:

_Enabled vs disabled in Chrome_
![Screen Shot 2022-12-08 at 12 46 27 PM](https://user-images.githubusercontent.com/25211650/206525797-f365383c-0f46-4b45-9d38-08824d8dba35.png)

Safari & Firefox do not offer any visual queues an element is disabled at this time. This can cause 

## The solution

Create USWDS styles for disabled forms to add visual context for users.

Text Input Disabled 
![image](https://user-images.githubusercontent.com/25211650/206555349-de064a28-12a5-4059-8a48-99dd7296896a.png)

## Affected components

All components within the `Forms Inputs` category on Storybook as well as Button

- [x] Character Count
- [x] Checkbox _(Disabled Styles existed previously)_ 
- [ ] Checklist _(Didn't seem necessary. Not a form input)_ 
- [x] Combo Box
- [x] Date Picker
- [x] Date Range Picker
- [x] File Input
- [x] Text Input Mask
- [x] Input Prefix or Suffix _(Work completed in #5004)_ 
- [x] Text Input _(Work completed in #5004)_
- [x] Memorable Date
- [x] Radio _(Disabled Styles existed previously)_  
- [x] Range
- [x] Time Picker _(Storybook seems to be having trouble rendering this one)_   

## Updates made
- All form controls and buttons have matching `disabled` and `aria-disabled` states
    - Every disabled element and its overlaid text meets WCAG AA contrast:
        - Disabled backgrounds are defined with **`$theme-color-disabled` ("gray-20")**
        - Text on default backgrounds are defined with the new variable: **`$theme-color-disabled-text` (”gray-50”)**
        - Text on reverse backgrounds are defined with the new variable: **`$theme-color-disabled-text-reverse` ("gray-40")**
        - Text on disabled backgrounds are defined with the new variable: **`$theme-color-text-on-disabled` ("gray-70")**
    - Disabled styles now receive `cursor: not-allowed`
    - Disabled styles do not show hover states
    - Removed all disabled classes (This only applies to user-added disabled styles. Dynamically-added styles are still allowed)
    - Dynamically-added disabled styles are triggered for both `disabled` and `aria-disabled`
- All disabled states are consistently demonstrated in Storybook
    - A new `Disabled Forms` story has been added that shows all disabled form elements on one page
    - All `disabled`/`aria-disabled` controls in storybook are consistent
    - For buttons, `aria-disabled` is now included in the “show all states” control, which is displayed by default
- Input prefix/suffix: removed need for a disabled class
    - The CSS was re-written to absolutely position the icons instead of relying on disabled parent class
    - Removed `index.js` and related test files because they were no longer needed (the JS moved focus to the disabled group, which no longer exists
- Removed test pattern twig for `usa-select—multiple`; can rely on standard twig instead

## New `disabled` utility mixin
The new `u-disabled` mixin overrides `background-color` and cursor styles of elements to display as disabled. Includes overrides for `hover`, `active`, and `focus` states.

Users can pass in `$bg-color` & `$text-color` parameters as necessary to add custom colors if needed.

An example of this can be found for `usa-button--outline` variants to ensure color compliance

```scss
@include u-disabled("transparent", $theme-color-disabled-text);
```

## New Disabled Color Tokens
| Token Name | Color |
| ------------- | ------------- |
| `$theme-color-disabled-text`  | "gray-50"  |
| `$theme-color-disabled-text-reverse` | "gray-40"  |
| `$theme-color-text-on-disabled` | "gray-70" |


## Preview
[Disabled Forms →](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/cm-disabled-forms-styles/?path=/story/patterns-forms--disabled-form-elements)

## To Test

- Visit each of the form components
- View disabled variant or turn on `disabled` and/or `aria-disabled`
- You may need to refreseh the page in order for the components javascript to initialize and the view to update
- Confirm:
   - `disabled` elements have a grey background and a `not-allowed` cursor on hover
   - `aria-disabled` styles mimic `disabled styles` _(Components will still be accessible if not `disabled`)_ 
   - Both style work together and add relative attribute to html code

---

### Notes & Other Considerations

After discussion, we have decided to allow disabled classes that style elements that:

- Do not receive the disabled attribute
- Are dynamically placed
   This will allow us to correctly style elements as disabled without putting the burden on the user.

Further findings from my disabled landscape analysis can be found at #5117 

⚠️ We should consider adding new disabled tokens to [theme color tokens](https://designsystem.digital.gov/design-tokens/color/theme-tokens/) and create a page for the disabled utility mixin in [Utilities](https://designsystem.digital.gov/utilities/)

Moving forward, we should also create guidance on our website about accessibility and disabled form elements to assist users in understanding when it is and is not appropriate to use these new styles and mixin

Before you hit Submit, make sure you’ve done whichever of these applies to you:
